File: Problems/Algorithms/530. Minimum Absolute Difference in BST/MinAbsoluteDifferenceBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int res = 100000;
    private int pre = -100000;

    public int getMinimumDifference(TreeNode root) {
        if (root.left != null) {
            getMinimumDifference(root.left);
        }

        res = Math.min(res, root.val - pre);
        pre = root.val;

        if (root.right != null) {
            getMinimumDifference(root.right);
        }

        return res;
    }
}

File: Problems/Algorithms/530. Minimum Absolute Difference in BST/min_absolute_difference_bst.py
#!/usr/bin/env python3

from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        node_values = []

        def dfs(node):
            if node is None:
                return
            node_values.append(node.val)
            dfs(node.left)
            dfs(node.right)
        
        dfs(root)

        node_values.sort()
        min_difference = 1e9
        for i in range(1, len(node_values)):
            min_difference = min(min_difference, node_values[i] - node_values[i-1])

        return min_difference
